File: src/main/GO/leetcode/lists/hot100/LeetCode85.go
package hot100

//DP： 打表后进行记忆化搜索，辅助以搜索空间
func maximalRectangle(matrix [][]byte) int {
	if len(matrix) == 0 {
		return 0
	}
	row, col := len(matrix), len(matrix[0])
	dp := make([][]int, row)
	for i := 0; i < row; i++ {
		dp[i] = make([]int, col)
	}
	maxArea := 0
	for i := 0; i < row; i++ {
		for j := 0; j < col; j++ {
			if i == 0 {
				if matrix[i][j] == '1' {
					dp[i][j] = 1
				}
			} else {
				if matrix[i][j] == '1' {
					dp[i][j] = dp[i-1][j] + 1
				}
			}
			min := dp[i][j]
			for k := j; k >= 0; k-- {
				if min == 0 {
					break
				}
				if dp[i][k] < min {
					min = dp[i][k]
				}
				maxArea = Max(maxArea, min*(j-k+1))
			}
		}
	}
	return maxArea
}

File: java/src/leetcode/problems/LeetCode85.java
package leetcode.problems;

/**
 * @author wujingxinit@outlook.com
 * @date 2024/4/11 18:52
 */
public class LeetCode85 {

    public int maximalRectangle(char[][] matrix) {
        int n = matrix[0].length;
        int max = 0;
        int[] h = new int[n];
        for (char[] row : matrix) {
            for (int j = 0; j < n; j++) {
                if (row[j] == '1') h[j]++;
                else h[j] = 0;
            }
            max = Math.max(max, new LeetCode84().largestRectangleArea(h));
        }
        return max;
    }
}
